### DISCUSSION TITLE
Assist reporting Internal Compiler Errors upstream

### DISCUSSION THREAD

**Discussion by tschwinge** [compiler-explorer/compiler-explorer#2633] [Ideas]
An idea that I just had, in context of GCC, but should likewise apply to any kind of compiler.  If a user's code triggers an Internal Compiler Error (ICE), that's -- by definition -- always a bug in the compiler, and the respective compiler's upstream developers should be interested in hearing about this, to fix their implementation.  Could CE help by offering an easy way to submit the "offending" code, compilation flags, etc. upstream?  (If the respective upstream does like getting such reports.)

CE would need to have logic to detect an ICE (in particular to differentiate it from a normal failed compilation due to invalid source code, compilation flags, etc.  (GCC: abnormal exit status, and _internal compiler error_ string in output, I suppose.)

CE would need to know where/how to report a bug upstream.  (Bugzilla, GitHub, etc.)

GCC has a command-line flag `-freport-bug` to _Collect and dump debug information into a temporary file if an internal compiler error (ICE) occurs_, which should be handy.

---

GCC examples:

  - <https://godbolt.org/z/Kx3bY8sd7> (already known with GCC upstream, I'm to fix that one) ;-)
  - <https://godbolt.org/z/4WWcr1sTM> (already known with GCC/Rust variant, <https://github.com/Rust-GCC/gccrs/issues/382>).

---

I have not yet considered how this relates to users' data privacy.

I have not yet considered the danger of DoS against upstream bug tracker.

If there's concern regarding these, then CE could still do the detection/collecting work, but the submission process would be a manual step for the user to do?

**Comment by mattgodbolt** [comment]
Great idea: We've talked about this a bunch of times. You identify some of the issues. We've looked at (and are still considering using) something like backtrace.io to do this (we generously have a license).

The issues are, like you say:

* Privacy. Though we have a flag in the settings "I'm prepared to let you use my code" and can extend it (with appropriate warning, policy updates)
* DoSing the upstream. Something like backtrace helps here as it can be the repo for bugs, and de-dupe (maybe).
* Tracking old versions - we'll need to only apply this to "actively supported" versions of compilers.
* Ignoring/tracking very carefully "trunk" builds - if there's a fleeting bug in the trunk builds then we don't want to spam with "already fixed".
* Upstream bug reporting for the compilers are very tightly controlled, with user account generation tricky. We'd need to liaise with the various compiler folks to either have a "Compiler Explorer" account, or have a way to log in some other way.
